warp10/src/main/java/io/warp10/script/functions/POLYFUNC.java

```diff
@@ -0,0 +1,171 @@
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.script.functions;
+
+import io.warp10.continuum.gts.GTSDecoder;
+import io.warp10.continuum.gts.GTSEncoder;
+import io.warp10.continuum.gts.GTSHelper;
+import io.warp10.continuum.gts.GeoTimeSerie;
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.StackUtils;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptLib;
+import io.warp10.script.WarpScriptMapperFunction;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.commons.math3.analysis.polynomials.PolynomialFunction;
+import org.apache.kafka.common.config.ConfigDef.Type;
+
+/**
+ * Build a function which will compute a polynomial function
+ */
+public class POLYFUNC extends NamedWarpScriptFunction implements WarpScriptStackFunction, WarpScriptMapperFunction {
+
+  private final PolynomialFunction func;
+
+  public static class Builder extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+    public Builder(String name) {
+      super(name);
+    }
+
+    @Override
+    public Object apply(WarpScriptStack stack) throws WarpScriptException {
+      Object top = stack.pop();
+
+      if (!(top instanceof List)) {
+        throw new WarpScriptException(getName() + " expects a list of polynomial coefficients on top of the stack.");
+      }
+
+      List<Object> l = (List<Object>) top;
+
+      double[] coeffs = new double[l.size()];
+
+      int i = 0;
+
+      for (Object o: l) {
+        if (!(o instanceof Number)) {
+          throw new WarpScriptException(getName() + " expects polynomial coefficients to be numerical.");
+        }
+        coeffs[i++] = ((Number) o).doubleValue();
+      }
+
+      stack.push(new POLYFUNC(getName(), coeffs));
+
+      return stack;
+    }
+  }
+
+  public POLYFUNC(String name, double[] coeffs) {
+    super(name);
+    this.func = new PolynomialFunction(coeffs);
+  }
+
+  @Override
+  public Object apply(WarpScriptStack stack) throws WarpScriptException {
+    Object top = stack.pop();
+
+    if (top instanceof List) {
+      List<Object> list = new ArrayList<Object>(((List) top).size());
+      for (Object elt: (List) top) {
+        if (!(elt instanceof Number)) {
+          throw new WarpScriptException(getName() + " can only be applied to numerical values.");
+        }
+        list.add(this.func.value(((Number) elt).doubleValue()));
+      }
+      stack.push(list);
+    } else if (top instanceof GeoTimeSerie) {
```

Comment:
To be coherent with POLYFIT, this function may also accept a `useBucketId` argument and raise an error if the argument below is not a bucketized numerical GTS.

Reply:
Nope, the function is created by POLYFIT based on some arguments, but there is no reason to constraint its use.

Reply:
The function would just have an additional optional argument. It's use would not be constrained.

Reply:
yes that is why I don't understand why the function created by POLYFUNC should not also have that optional parameter when applied on bucketized GTS. It doesn't interfere with when this created function is applied on other input types.

Reply:
Ok, got what you meant! Sorry for my slow brain today!